arch/risc-v/src/common/riscv_cpuindex.c

```diff
@@ -48,10 +50,14 @@
  *
  ****************************************************************************/
 
-int up_cpu_index(void)
+uintptr_t riscv_cpuindex(void)
 {
-  int mhartid;
+  return READ_CSR(mhartid);
+}
 
-  asm volatile ("csrr %0, mhartid": "=r" (mhartid));
-  return mhartid;
+#ifdef CONFIG_SMP
+int up_cpu_index(void)
+{
+  return (int)riscv_cpuindex;
```

Comment:
This should just become a tail call and is not dependent on preprocessor substitution, would this solution suffice ?